### PR TITLE
[Relay][Frontend] Keras Support

### PR DESCRIPTION
**Pull request by Huyuwei**
Keras -> relay frontend migrated from nnvm with one main change that SymbolTable is replaced by ExprTable.

@kazum @thefiddler welcome to review.

### CODE REVIEW

File: python/tvm/relay/frontend/common.py [hunk 1]
```diff
@@ -204,3 +203,27 @@ def get_bool(self, key, default=RequiredAttr()):
         if isinstance(default, RequiredAttr):
             raise AttributeError("Required attribute {} not found.".format(key))
         return default
+
+
+class ExprTable(object):
+    """Table storing relay exprs by names."""
```
Comment (jroesch) [inline python/tvm/relay/frontend/common.py]:
```suggestion
    """Table storing Relay expressions by names."""
```

File: python/tvm/relay/frontend/keras.py [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+
+
+def _check_unsupported_layers(model):
+    for layer in model.layers:
+        if type(layer).__name__ not in _convert_map:
+            raise ValueError("Keras layer {} not supported.".format(type(layer).__name__))
+
+
+def keras_op_to_relay(inexpr, keras_layer, outname, etab):
+    """Convert keras layer to relay expr, and update etab.
```
Comment (jroesch) [inline python/tvm/relay/frontend/keras.py]:
```suggestion
    """Convert a Keras layer to a Relay expression and update the expression table."""
```

File: tests/python/relay/frontend/keras/test_forward.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    test_forward_pool()
+    test_forward_conv()
+    test_forward_upsample()
+    test_forward_reshape()
+    test_forward_crop()
+    test_forward_multi_inputs()
+    test_forward_multi_outputs()
+    test_forward_reuse_layers()
+    test_forward_rnn()
+
```
Comment (jroesch) [inline tests/python/relay/frontend/keras/test_forward.py]: Can we remove this line.

File: python/tvm/relay/frontend/keras.py [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+        The input relay expr(s)
+
+    keras_layer : keras.layers
+        The keras layer to be converted
+
+    outname : str
+        Name of the output relay expr
+
+    etab : relay.frontend.common.ExprTable
+        The global expr table to be updated
```
Comment (jroesch) [inline python/tvm/relay/frontend/keras.py]:
```suggestion
        The global expression table to be updated.
```

File: python/tvm/relay/frontend/keras.py [hunk 5]
```diff
@@ -0,0 +1,702 @@
...
+    Parameters
+    ----------
+    inexpr : relay.expr.Expr or a list of it
+        The input relay expr(s)
+
+    keras_layer : keras.layers
+        The keras layer to be converted
+
+    outname : str
+        Name of the output relay expr
```
Comment (jroesch) [inline python/tvm/relay/frontend/keras.py]:
```suggestion
        Name of the output Relay expression.
```

File: python/tvm/relay/frontend/keras.py [hunk 6]
```diff
@@ -0,0 +1,702 @@
...
+def keras_op_to_relay(inexpr, keras_layer, outname, etab):
+    """Convert keras layer to relay expr, and update etab.
+
+    Parameters
+    ----------
+    inexpr : relay.expr.Expr or a list of it
+        The input relay expr(s)
+
+    keras_layer : keras.layers
+        The keras layer to be converted
```
Comment (jroesch) [inline python/tvm/relay/frontend/keras.py]:
```suggestion
        The Keras layer to be converted.
```

File: python/tvm/relay/frontend/keras.py [hunk 7]
```diff
@@ -0,0 +1,702 @@
...
+            raise ValueError("Keras layer {} not supported.".format(type(layer).__name__))
+
+
+def keras_op_to_relay(inexpr, keras_layer, outname, etab):
+    """Convert keras layer to relay expr, and update etab.
+
+    Parameters
+    ----------
+    inexpr : relay.expr.Expr or a list of it
+        The input relay expr(s)
```
Comment (jroesch) [inline python/tvm/relay/frontend/keras.py]:
```suggestion
        The input Relay expression(s).
```

File: Jenkinsfile [hunk 8]
```diff
@@ -236,6 +236,17 @@ stage('Integration Test') {
       }
     }
   },
+  'frontend: GPU': {
```
Comment (Huyuwei) [inline Jenkinsfile]: @tqchen please check the Jenkinsfile change to add relay frontend tests. Do we have to merge it so that it can take effect?

Should we put all relay frontend tests in tests/python/frontend? I see mxnet tests in both tests/python/frontend and tests/python/relay/frontend

**Comment by tqchen** [comment]
@jroesch please take another look, @kazum please take some time to take a look at the PR, I plan to bring this in by the end of the week if there is no further comments

File: python/tvm/relay/frontend/keras.py [hunk 9]
```diff
@@ -0,0 +1,702 @@
...
+    # we insert a separate pad operator
+    elif keras_layer.padding == 'same':
+        in_h = keras_layer.input_shape[1]
+        in_w = keras_layer.input_shape[2]
+        pad_t, pad_b = _get_pad_pair(in_h, dilated_kernel_h, stride_h)
+        pad_l, pad_r = _get_pad_pair(in_w, dilated_kernel_w, stride_w)
+        if pad_t == pad_b and pad_l == pad_r:
+            params['padding'] = (pad_t, pad_l)
+        else:
+            inexpr = _op.nn.pad(data=inexpr, pad_width=(
```
[the remark this answers is not in the document]
Comment (Huyuwei) [reply python/tvm/relay/frontend/keras.py]: Sure, this is doable. It would be great if someone is interested in contributing this.

[the remark this answers is not in the document]
Comment (FrozenGene) [reply python/tvm/relay/frontend/keras.py]: I am worry the compatibility too. NNVM / our relay FE uses 2D padding.

[the remark this answers is not in the document]
Comment (FrozenGene) [reply python/tvm/relay/frontend/keras.py]: Moreever, x86 backend assume padding is 2D too, which assume only pad_h and pad_w. So this is great, but seem that many places we should be careful.

**Comment by tqchen** [comment]
@Huyuwei there is an testcase error in master, which could be caused by our recent change of convention to const  http://ci.tvm.ai:8080/blue/organizations/jenkins/tvm/detail/master/577/pipeline https://github.com/dmlc/tvm/pull/2349
can you make a quick fix?

**Comment by Huyuwei** [comment]
@tqchen fixed by #2376
